File: pkg/runtime/runtimes/ts-cjs/pulserpc/tests/test_validation.js
/**
 * Tests for validation functions (CommonJS variant)
 */

const { strict: assert } = require("assert");
const {
  validateString,
  validateInt,
  validateFloat,
  validateBool,
  validateArray,
  validateMap,
  validateEnum,
  validateStruct,
  validateType,
} = require("../validation");

function testValidateStringSuccess() {
  assert.deepEqual(validateString("hello"), []);
  assert.deepEqual(validateString(""), []);
  console.log("✓ testValidateStringSuccess");
}

function testValidateStringFailure() {
  const errs = validateString(123);
  assert.equal(errs.length, 1);
  assert.ok(errs[0].message.includes("Expected string"));
  assert.equal(errs[0].path, "");

  const errs2 = validateString(null);
  assert.equal(errs2.length, 1);
  assert.ok(errs2[0].message.includes("Expected string"));
  console.log("✓ testValidateStringFailure");
}

function testValidateIntSuccess() {
  assert.deepEqual(validateInt(0), []);
  assert.deepEqual(validateInt(42), []);
  assert.deepEqual(validateInt(-100), []);
  assert.deepEqual(validateInt(5.0), []);
  console.log("✓ testValidateIntSuccess");
}

function testValidateIntFailure() {
  const errs = validateInt("123");
  assert.equal(errs.length, 1);
  assert.ok(errs[0].message.includes("Expected number for int"));

  const errs2 = validateInt(3.14);
  assert.equal(errs2.length, 1);
  assert.ok(errs2[0].message.includes("Expected integer"));

  assert.equal(validateInt(5.1).length, 1);
  console.log("✓ testValidateIntFailure");
}

function testValidateFloatSuccess() {
  assert.deepEqual(validateFloat(3.14), []);
  assert.deepEqual(validateFloat(42), []);
  assert.deepEqual(validateFloat(-1.5), []);
  console.log("✓ testValidateFloatSuccess");
}

function testValidateFloatFailure() {
  const errs = validateFloat("3.14");
  assert.equal(errs.length, 1);
  assert.ok(errs[0].message.includes("Expected number for float"));

  assert.equal(validateFloat(null).length, 1);
  console.log("✓ testValidateFloatFailure");
}

function testValidateBoolSuccess() {
  assert.deepEqual(validateBool(true), []);
  assert.deepEqual(validateBool(false), []);
  console.log("✓ testValidateBoolSuccess");
}

function testValidateBoolFailure() {
  assert.equal(validateBool(1).length, 1);
  assert.equal(validateBool("true").length, 1);
  console.log("✓ testValidateBoolFailure");
}

function testValidateArraySuccess() {
  const elementValidator = (v, p) => validateString(v, p);
  assert.deepEqual(validateArray(["a", "b", "c"], elementValidator), []);
  assert.deepEqual(validateArray([], elementValidator), []);
  console.log("✓ testValidateArraySuccess");
}

function testValidateArrayWrongType() {
  const elementValidator = (v, p) => validateString(v, p);
  const errs = validateArray("not a list", elementValidator);
  assert.equal(errs.length, 1);
  assert.ok(errs[0].message.includes("Expected array"));
  console.log("✓ testValidateArrayWrongType");
}

function testValidateArrayElementValidationFails() {
  const elementValidator = (v, p) => validateString(v, p);
  const errs = validateArray(["a", 123, "c"], elementValidator);
  assert.equal(errs.length, 1);
  assert.ok(errs[0].path.includes("[1]"));
  assert.ok(errs[0].message.includes("Expected string"));
  console.log("✓ testValidateArrayElementValidationFails");
}

function testValidateArrayPathTracking() {
  const elementValidator = (v, p) => validateString(v, p);
  const errs = validateArray(["a", "b", 42], elementValidator, ".items");
  assert.equal(errs.length, 1);
  assert.equal(errs[0].path, ".items[2]");
  console.log("✓ testValidateArrayPathTracking");
}

function testValidateMapSuccess() {
  const valueValidator = (v, p) => validateInt(v, p);
  assert.deepEqual(validateMap({ a: 1, b: 2 }, valueValidator), []);
  assert.deepEqual(validateMap({}, valueValidator), []);
  console.log("✓ testValidateMapSuccess");
}

function testValidateMapWrongType() {
  const valueValidator = (v, p) => validateInt(v, p);
  const errs1 = validateMap("not a dict", valueValidator);
  assert.equal(errs1.length, 1);
  assert.ok(errs1[0].message.includes("Expected object for map"));

  const errs2 = validateMap([], valueValidator);
  assert.equal(errs2.length, 1);
  console.log("✓ testValidateMapWrongType");
}

function testValidateMapValueValidationFails() {
  const valueValidator = (v, p) => validateInt(v, p);
  const errs = validateMap({ a: "not an int" }, valueValidator);
  assert.equal(errs.length, 1);
  assert.ok(errs[0].path.includes("[a]"));
  console.log("✓ testValidateMapValueValidationFails");
}

function testValidateEnumSuccess() {
  assert.deepEqual(validateEnum("kindle", "Platform", ["kindle", "nook"]), []);
  assert.deepEqual(validateEnum("nook", "Platform", ["kindle", "nook"]), []);
  console.log("✓ testValidateEnumSuccess");
}

function testValidateEnumWrongType() {
  const errs = validateEnum(123, "Platform", ["kindle", "nook"]);
  assert.equal(errs.length, 1);
  assert.ok(errs[0].message.includes("Expected string for enum"));
  console.log("✓ testValidateEnumWrongType");
}

function testValidateEnumInvalidValue() {
  const errs = validateEnum("invalid", "Platform", ["kindle", "nook"]);
  assert.equal(errs.length, 1);
  assert.ok(errs[0].message.includes("Invalid value for enum"));
  console.log("✓ testValidateEnumInvalidValue");
}

function testValidateStructSuccess() {
  const allStructs = {
    User: {
      fields: [
        { name: "id", type: { builtIn: "string" }, optional: false },
        { name: "name", type: { builtIn: "string" }, optional: false },
      ],
    },
  };
  const allEnums = {};

  const errs = validateStruct(
    { id: "123", name: "Alice" },
    "User",
    allStructs["User"],
    allStructs,
    allEnums
  );
  assert.deepEqual(errs, []);
  console.log("✓ testValidateStructSuccess");
}

function testValidateStructMissingRequiredField() {
  const allStructs = {
    User: {
      fields: [{ name: "id", type: { builtIn: "string" }, optional: false }],
    },
  };
  const allEnums = {};

  const errs = validateStruct({}, "User", allStructs["User"], allStructs, allEnums);
  assert.equal(errs.length, 1);
  assert.ok(errs[0].message.includes("Missing required field"));
  assert.equal(errs[0].path, ".id");
  console.log("✓ testValidateStructMissingRequiredField");
}

function testValidateStructOptionalField() {
  const allStructs = {
    User: {
      fields: [
        { name: "id", type: { builtIn: "string" }, optional: false },
        { name: "email", type: { builtIn: "string" }, optional: true },
      ],
    },
  };
  const allEnums = {};

  assert.deepEqual(validateStruct({ id: "123" }, "User", allStructs["User"], allStructs, allEnums), []);
  assert.deepEqual(validateStruct({ id: "123", email: "alice@example.com" }, "User", allStructs["User"], allStructs, allEnums), []);
  console.log("✓ testValidateStructOptionalField");
}

function testValidateStructWithExtends() {
  const allStructs = {
    Base: {
      fields: [{ name: "id", type: { builtIn: "string" }, optional: false }],
    },
    User: {
      extends: "Base",
      fields: [{ name: "name", type: { builtIn: "string" }, optional: false }],
    },
  };
  const allEnums = {};

  assert.deepEqual(
    validateStruct({ id: "123", name: "Alice" }, "User", allStructs["User"], allStructs, allEnums),
    []
  );

  const errs = validateStruct({ name: "Alice" }, "User", allStructs["User"], allStructs, allEnums);
  assert.equal(errs.length, 1);
  assert.ok(errs[0].message.includes("Missing required field"));
  assert.equal(errs[0].path, ".id");
  console.log("✓ testValidateStructWithExtends");
}

function testValidateStructCollectsMultipleErrors() {
  const allStructs = {
    Person: {
      fields: [
        { name: "username", type: { builtIn: "string" } },
        { name: "age", type: { builtIn: "int" } },
        { name: "email", type: { builtIn: "string" } },
      ],
    },
  };
  const allEnums = {};

  const errs = validateStruct(
    { username: "alice", age: "not-a-number", email: 42 },
    "Person",
    allStructs["Person"],
    allStructs,
    allEnums
  );
  assert.equal(errs.length, 2);
  assert.equal(errs[0].path, ".age");
  assert.equal(errs[1].path, ".email");
  console.log("✓ testValidateStructCollectsMultipleErrors");
}

function testValidateTypeString() {
  const allStructs = {};
  const allEnums = {};
  assert.deepEqual(validateType("hello", { builtIn: "string" }, allStructs, allEnums), []);
  console.log("✓ testValidateTypeString");
}

function testValidateTypeOptionalNone() {
  const allStructs = {};
  const allEnums = {};

  assert.deepEqual(validateType(null, { builtIn: "string" }, allStructs, allEnums, true), []);

  const errs1 = validateType(undefined, { builtIn: "string" }, allStructs, allEnums, true);
  assert.equal(errs1.length, 1);
  assert.ok(errs1[0].message.includes("cannot be undefined"));

  const errs2 = validateType(null, { builtIn: "string" }, allStructs, allEnums, false);
  assert.equal(errs2.length, 1);
  assert.ok(errs2[0].message.includes("cannot be null"));
  console.log("✓ testValidateTypeOptionalNone");
}

function testValidateTypeArray() {
  const allStructs = {};
  const allEnums = {};
  const typeDef = { array: { builtIn: "string" } };

  assert.deepEqual(validateType(["a", "b"], typeDef, allStructs, allEnums), []);

  const errs = validateType(["a", 123], typeDef, allStructs, allEnums);
  assert.equal(errs.length, 1);
  assert.ok(errs[0].path.includes("[1]"));
  console.log("✓ testValidateTypeArray");
}

function testValidateTypeMap() {
  const allStructs = {};
  const allEnums = {};
  const typeDef = { mapValue: { builtIn: "int" } };

  assert.deepEqual(validateType({ a: 1, b: 2 }, typeDef, allStructs, allEnums), []);

  const errs = validateType({ a: "not int" }, typeDef, allStructs, allEnums);
  assert.equal(errs.length, 1);
  assert.ok(errs[0].path.includes("[a]"));
  console.log("✓ testValidateTypeMap");
}

function testValidateNestedStructInArray() {
  const allStructs = {
    Child: {
      fields: [
        { name: "name", type: { builtIn: "string" } },
        { name: "age", type: { builtIn: "int" } },
      ],
    },
    Person: {
      fields: [
        { name: "name", type: { builtIn: "string" } },
        { name: "children", type: { array: { userDefined: "Child" } } },
      ],
    },
  };
  const allEnums = {};

  assert.deepEqual(
    validateType(
      { name: "Alice", children: [{ name: "Bob", age: 10 }, { name: "Charlie", age: 12 }] },
      { userDefined: "Person" },
      allStructs,
      allEnums
    ),
    []
  );

  const errs = validateType(
    { name: "Alice", children: [{ name: "Bob", age: 10 }, { name: "Charlie", age: "twelve" }] },
    { userDefined: "Person" },
    allStructs,
    allEnums
  );
  assert.equal(errs.length, 1);
  assert.equal(errs[0].path, ".children[1].age");
  console.log("✓ testValidateNestedStructInArray");
}

testValidateStringSuccess();
testValidateStringFailure();
testValidateIntSuccess();
testValidateIntFailure();
testValidateFloatSuccess();
testValidateFloatFailure();
testValidateBoolSuccess();
testValidateBoolFailure();
testValidateArraySuccess();
testValidateArrayWrongType();
testValidateArrayElementValidationFails();
testValidateArrayPathTracking();
testValidateMapSuccess();
testValidateMapWrongType();
testValidateMapValueValidationFails();
testValidateEnumSuccess();
testValidateEnumWrongType();
testValidateEnumInvalidValue();
testValidateStructSuccess();
testValidateStructMissingRequiredField();
testValidateStructOptionalField();
testValidateStructWithExtends();
testValidateStructCollectsMultipleErrors();
testValidateTypeString();
testValidateTypeOptionalNone();
testValidateTypeArray();
testValidateTypeMap();
testValidateNestedStructInArray();
console.log("\nAll validation tests passed!");

File: pkg/runtime/runtimes/python/tests/test_validation.py
"""Tests for validation functions"""

import pytest
from pulserpc import (
    validate_string,
    validate_int,
    validate_float,
    validate_bool,
    validate_array,
    validate_map,
    validate_enum,
    validate_struct,
    validate_type,
)


class TestBuiltInTypes:
    """Test built-in type validators"""
    
    def test_validate_string_success(self):
        validate_string("hello")
        validate_string("")
    
    def test_validate_string_failure(self):
        errs = validate_string(123)
        assert len(errs) == 1
        assert "Expected string" in errs[0].message

        errs2 = validate_string(None)
        assert len(errs2) == 1
        assert "Expected string" in errs2[0].message
    
    def test_validate_int_success(self):
        validate_int(0)
        validate_int(42)
        validate_int(-100)
    
    def test_validate_int_failure(self):
        errs = validate_int("123")
        assert len(errs) == 1
        assert "Expected int" in errs[0].message

        errs2 = validate_int(3.14)
        assert len(errs2) == 1
        assert "Expected int" in errs2[0].message

        # Test bool rejection
        errs3 = validate_int(True)
        assert len(errs3) == 1
        assert "Expected int" in errs3[0].message

        # Test infinity
        errs4 = validate_int(float('inf'))
        assert len(errs4) == 1
        assert "Expected int" in errs4[0].message

        # Test NaN
        errs5 = validate_int(float('nan'))
        assert len(errs5) == 1
        assert "Expected int" in errs5[0].message
    
    def test_validate_float_success(self):
        validate_float(3.14)
        validate_float(42)  # int is acceptable
        validate_float(-1.5)
    
    def test_validate_float_failure(self):
        errs = validate_float("3.14")
        assert len(errs) == 1
        assert "Expected float" in errs[0].message

        errs2 = validate_float(None)
        assert len(errs2) == 1
        assert "Expected float" in errs2[0].message
    
    def test_validate_bool_success(self):
        validate_bool(True)
        validate_bool(False)
    
    def test_validate_bool_failure(self):
        errs = validate_bool(1)
        assert len(errs) == 1
        assert "Expected bool" in errs[0].message

        errs2 = validate_bool("true")
        assert len(errs2) == 1
        assert "Expected bool" in errs2[0].message


class TestArrayValidation:
    """Test array validation"""
    
    def test_validate_array_success(self):
        element_validator = lambda v, p: validate_string(v, p)
        errs = validate_array(["a", "b", "c"], element_validator)
        assert len(errs) == 0
        errs2 = validate_array([], element_validator)
        assert len(errs2) == 0
    
    def test_validate_array_wrong_type(self):
        element_validator = lambda v, p: validate_string(v, p)
        errs = validate_array("not a list", element_validator)
        assert len(errs) == 1
        assert "Expected list" in errs[0].message

        errs2 = validate_array({}, element_validator)
        assert len(errs2) == 1
        assert "Expected list" in errs2[0].message
    
    def test_validate_array_element_validation_fails(self):
        element_validator = lambda v, p: validate_string(v, p)
        errs = validate_array(["a", 123, "c"], element_validator)
        assert len(errs) == 1
        assert "[1]" in errs[0].path
        assert "Expected string" in errs[0].message


class TestMapValidation:
    """Test map validation"""
    
    def test_validate_map_success(self):
        value_validator = lambda v, p: validate_int(v, p)
        errs = validate_map({"a": 1, "b": 2}, value_validator)
        assert len(errs) == 0
        errs2 = validate_map({}, value_validator)
        assert len(errs2) == 0
    
    def test_validate_map_wrong_type(self):
        value_validator = lambda v, p: validate_int(v, p)
        errs = validate_map("not a dict", value_validator)
        assert len(errs) == 1
        assert "Expected dict" in errs[0].message

        errs2 = validate_map([], value_validator)
        assert len(errs2) == 1
        assert "Expected dict" in errs2[0].message
    
    def test_validate_map_non_string_key(self):
        value_validator = lambda v, p: validate_int(v, p)
        errs = validate_map({123: 1}, value_validator)
        assert len(errs) == 1
        assert "Map key must be string" in errs[0].message
    
    def test_validate_map_value_validation_fails(self):
        value_validator = lambda v, p: validate_int(v, p)
        errs = validate_map({"a": "not an int"}, value_validator)
        assert len(errs) == 1
        assert "[a]" in errs[0].path


class TestEnumValidation:
    """Test enum validation"""
    
    def test_validate_enum_success(self):
        errs = validate_enum("kindle", "Platform", ["kindle", "nook"])
        assert len(errs) == 0
        errs2 = validate_enum("nook", "Platform", ["kindle", "nook"])
        assert len(errs2) == 0
    
    def test_validate_enum_wrong_type(self):
        errs = validate_enum(123, "Platform", ["kindle", "nook"])
        assert len(errs) == 1
        assert "Expected string for enum" in errs[0].message
    
    def test_validate_enum_invalid_value(self):
        errs = validate_enum("invalid", "Platform", ["kindle", "nook"])
        assert len(errs) == 1
        assert "Invalid value for enum" in errs[0].message


class TestStructValidation:
    """Test struct validation"""
    
    def test_validate_struct_success(self):
        all_structs = {
            'User': {
                'fields': [
                    {'name': 'id', 'type': {'builtIn': 'string'}, 'optional': False},
                    {'name': 'name', 'type': {'builtIn': 'string'}, 'optional': False},
                ]
            }
        }
        all_enums = {}
        struct_def = all_structs['User']
        
        errs = validate_struct(
            {'id': '123', 'name': 'Alice'},
            'User',
            struct_def,
            all_structs,
            all_enums
        )
        assert len(errs) == 0
    
    def test_validate_struct_missing_required_field(self):
        all_structs = {
            'User': {
                'fields': [
                    {'name': 'id', 'type': {'builtIn': 'string'}, 'optional': False},
                ]
            }
        }
        all_enums = {}
        struct_def = all_structs['User']
        
        errs = validate_struct({}, 'User', struct_def, all_structs, all_enums)
        assert len(errs) == 1
        assert "Missing required field" in errs[0].message
    
    def test_validate_struct_optional_field(self):
        all_structs = {
            'User': {
                'fields': [
                    {'name': 'id', 'type': {'builtIn': 'string'}, 'optional': False},
                    {'name': 'email', 'type': {'builtIn': 'string'}, 'optional': True},
                ]
            }
        }
        all_enums = {}
        struct_def = all_structs['User']
        
        # Should work without optional field
        errs = validate_struct({'id': '123'}, 'User', struct_def, all_structs, all_enums)
        assert len(errs) == 0
        
        # Should work with optional field
        errs2 = validate_struct(
            {'id': '123', 'email': 'alice@example.com'},
            'User',
            struct_def,
            all_structs,
            all_enums
        )
        assert len(errs2) == 0
    
    def test_validate_struct_with_extends(self):
        all_structs = {
            'Base': {
                'fields': [
                    {'name': 'id', 'type': {'builtIn': 'string'}, 'optional': False},
                ]
            },
            'User': {
                'extends': 'Base',
                'fields': [
                    {'name': 'name', 'type': {'builtIn': 'string'}, 'optional': False},
                ]
            }
        }
        all_enums = {}
        struct_def = all_structs['User']
        
        # Should validate both parent and child fields
        errs = validate_struct(
            {'id': '123', 'name': 'Alice'},
            'User',
            struct_def,
            all_structs,
            all_enums
        )
        assert len(errs) == 0
        
        # Should fail if parent field missing
        errs2 = validate_struct({'name': 'Alice'}, 'User', struct_def, all_structs, all_enums)
        assert len(errs2) == 1
        assert "Missing required field" in errs2[0].message


class TestTypeValidation:
    """Test main validate_type function"""
    
    def test_validate_type_string(self):
        all_structs = {}
        all_enums = {}
        errs = validate_type("hello", {'builtIn': 'string'}, all_structs, all_enums)
        assert len(errs) == 0
    
    def test_validate_type_optional_none(self):
        all_structs = {}
        all_enums = {}
        
        # null is valid for optional fields
        errs = validate_type(None, {'builtIn': 'string'}, all_structs, all_enums, is_optional=True)
        assert len(errs) == 0
        
        # null is NOT valid for non-optional fields
        errs2 = validate_type(None, {'builtIn': 'string'}, all_structs, all_enums, is_optional=False)
        assert len(errs2) == 1
        assert "cannot be None" in errs2[0].message
    
    def test_validate_type_array(self):
        all_structs = {}
        all_enums = {}
        type_def = {'array': {'builtIn': 'string'}}
        
        errs = validate_type(["a", "b"], type_def, all_structs, all_enums)
        assert len(errs) == 0
        
        errs2 = validate_type(["a", 123], type_def, all_structs, all_enums)
        assert len(errs2) == 1
        assert "[1]" in errs2[0].path
    
    def test_validate_type_map(self):
        all_structs = {}
        all_enums = {}
        type_def = {'mapValue': {'builtIn': 'int'}}
        
        errs = validate_type({"a": 1, "b": 2}, type_def, all_structs, all_enums)
        assert len(errs) == 0
        
        errs2 = validate_type({"a": "not int"}, type_def, all_structs, all_enums)
        assert len(errs2) == 1
        assert "[a]" in errs2[0].path

